zerver/lib/url_decoding.py

```diff
@@ -24,3 +28,127 @@ def is_same_server_message_link(url: str) -> bool:
     ends_with_near_message_id = fragment_parts[-2] == "near" and fragment_parts[-1].isdigit()
 
     return category == "narrow" and section in {"channel", "dm"} and ends_with_near_message_id
+
+
+CHANNEL_SYNONYMS = {"stream": "channel", "streams": "channels"}
+
+OPERATOR_SYNONYMS = {
+    **CHANNEL_SYNONYMS,
+    # "pm-with:" was renamed to "dm:"
+    "pm-with": "dm",
+    # "group-pm-with:" was replaced with "dm-including:"
+    "group-pm-with": "dm-including",
+    "from": "sender",
+    DB_TOPIC_NAME: "topic",
+}
+
+
+def canonicalize_operator_synonyms(text: str) -> str:
+    text = text.lower()
+    if text in OPERATOR_SYNONYMS.values():
+        return text
+    if text in OPERATOR_SYNONYMS:
+        return OPERATOR_SYNONYMS[text]
+    return text
+
+
+def parse_recipient_slug(slug: str) -> tuple[Any, str] | None:
+    """
+    Parses operands formatted in slug containing object ID or IDs.
+    Typical of "channel" or private message operands.
+
+    Doesn't parse the legacy stream slug.
+        e.g. "stream-name"
+
+    Returns a tuple of parsed ids and the recipient info (channel name,
+    DM'ed users name, etc) or only `None` if the operand is invalid.
+        e.g.
+        - "12,13,14-group" -> ([12, 13, 14], "group")
+        - "89-Markl" -> (89, "Markl")
+        - "stream-name" -> None
+    """
+    try:
+        ids_string, suffix = slug.split("-", maxsplit=1)
+        ids = [int(id) for id in ids_string.split(",")]
+        return (ids if len(ids) > 1 else ids[0], suffix)
+    except ValueError:
+        # We primarily expect this to catch the legacy stream
+        # slug or broken URLs.
+        return None
+
+
+def hash_util_decode(string: str) -> str:
+    # Do the same decoding operation as shared internal_url.decodeHashComponent
+    # on the frontend.
+    return unquote(string.replace(".", "%"))
+
+
+def decode_narrow_operand(operator: str, operand: str) -> Any:
+    # These have the similar slug formatting for their operands which contains
+    # object ID(s).
+    if operator in ["dm-including", "dm", "sender", "channel"]:
+        result = parse_recipient_slug(operand)
+        return result[0] if isinstance(result, tuple) else ""
+
+    if operator == "near":
+        return int(operand) if operand.isdigit() else ""
+
+    operand = hash_util_decode(operand).strip()
+
+    return operand
+
+
+def parse_narrow_url(
+    narrow_url: str,
+) -> list[NarrowTerm] | None:
+    """
+    This is the web app version of `parse_narrow` in `hash_util.ts`. It
+    largely behaves the same way and has the same purpose: to parse a
+    narrow URL into a list of `NarrowTerm`.
+
+    The differences, other than some algorithms being done in a "Pythonic"
+    way, are:
+
+        - This doesn't convert a user-id-slug into a list of user emails;
+        it will instead parse it into a list of their IDs.
+
+        - Operands can be an actual list or int instead of a list or int as
+        a string (e.g., "12,14,15" or "93").
```

Comment:
I decided to drop mention of those first two things, since they are probably things we'll change about the web app version.